[feat] engines: add tavily (AI powered)

GenericMale opened the pull request:
## What does this PR do?

- Adds [Tavily](https://tavily.com) Engine
- Requires API key. Free Account allows 1000 requests per month.
- Support for general search and news.
- By default fills Answer Box with AI response.
- If images are turned on (`include_images: true`), the infobox is used instead.
- Supports time range
- Allows filtering of analyzed domains (`include_domains` & `exclude_domains`)

## How to test this PR locally?

- Create a (free) account on tavily.com or hmu for a key
- Enable engine in settings
- `!tav which engines does searxng support?`

## Examples

![20250121_204524](https://github.com/user-attachments/assets/e6dfd743-a5f3-49ce-b874-8b3b201ea799)
*Default Setting: Results & Answer*

![20250121_205214](https://github.com/user-attachments/assets/778c6570-da19-4d88-bfdc-8c74c4955410)
*`include_images: true`: Results & Infobox*

GenericMale committed:
[feat] new engine: tavily.com

return42 committed:
[mod] add doc to tavily and slightly improve the engine

- Config options like ``search_type`` renamed to follow the upstream
  API (``topic``).
- Default ``max_results`` is set to 5
- use image description if one exists
- add an init function to check engine's settings
- settings example: additional category 'ai'

To review the added documentation of this path::

    make docs.live

and jump to: http://0.0.0.0:8000/dev/engines/online/tavily.html

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 committed:
[mod] update intersphinx links and add missing CATEGORY_GROUPS

To get translations for, the missed CATEGORY_GROUPS has been added:

- ai
- movies
- translate
- wikimedia

Signed-off-by: Markus Heiser <markus.heiser@darmarit.de>

return42 reviewed (approved):
I pushed two commits on top of this branch, for details, take a look at the commit messages.

Please run another test with my changes ... to see if the engine still meets your expectations after my changes.

Thanks a lot for this PR .. it was really an excellent work and a joy to review :+1:

GenericMale commented on searx/engines/tavily.py:
```diff
@@ -1,89 +1,240 @@
-    params['data'] = dumps(data)
+
+    if include_images:
+        data["include_image_descriptions"] = True
```
could we make this optional (engine config) or just disable it? increases the processing time and i feel the generated descriptions aren't really useful for our use case.

GenericMale commented on searx/engines/tavily.py:
```diff
@@ -1,89 +1,240 @@
-                'url': result['url'],
-                'content': result['content'],
-                'publishedDate': _parse_date(result.get('published_date')),
+                "title": f"[{gettext('ai')}] {result['title']}",
```
can we move the [ai] prefix to the content instead? the title returned is the original page title afaict. only the content is a summarized version

GenericMale reviewed (commented):
thanks for the improvements, just two minor remarks

GenericMale committed:
[mod] small tavily engine changes
- add include_image_descriptions & include_answer to engine settings
- move [ai] prefix in results from title to content content
- minor doc fixes

return42 commented on searx/engines/tavily.py:
```diff
@@ -6,18 +6,18 @@
    please consult `API error codes`_.
 
-Tavily_ search API (AI engine).  This engine implements the REST API
+Tavily_ search API (AI engine). This engine implements the REST API
```
The double space after the dot of sentence end is intended.

Typographic tools use the two spaces after the end of a sentence to distinguish from the period of an abbreviation (by instance some editors adjust their automated line breaks accordingly).

return42 commented on searx/engines/tavily.py:
```diff
@@ -0,0 +1,246 @@
+            if not content:
+                result["content"] = img_caption
+            else:
+                result["content"] = f"{content}<br/>{img_caption}"
```
Do not use HTML markups in the results .. "separation of content structure and layout" 

- "structure" is given by the results their [types (I'm working on)](https://github.com/searxng/searxng/pull/4183) and their order.

- "layout" is done by the HTML templates & CSS

.. or in more practical terms: remember, we have also a JSON output. 

But for now leave it as is .. I will come back and review this PR next week / BTW thanks for your additions and corrections :+1:

return42 commented on searx/engines/tavily.py:
```diff
@@ -178,33 +185,32 @@ def response(resp):
-            img_src = img_list[0]["url"]
-            img_caption = gettext("Image caption") + ": " + img_list[0]["description"]
+            result["img_src"] = img_list[0]["url"]
+            img_caption = f"<i>{gettext('Image caption')}</i>: {img_list[0]['description']}"
```
Not using an f-string here was intended -->  [we can't combine `gettext` and f-string: it won't be detected by babel](https://github.com/searxng/searxng/pull/3413#issuecomment-2068081400).

GenericMale replied:
ah, reverted this and also removed the other f-strings containing a gettext

GenericMale committed:
[mod] tavily engine: revert double punctuation & remove f-stringed gettexts
